Library/Homebrew/cask/lib/hbc/quarantine.rb

```diff
@@ -0,0 +1,67 @@
+require "development_tools"
+module Hbc
+  module Quarantine
+    module_function
+
+    QUARANTINE_ATTRIBUTE = "com.apple.quarantine".freeze
+
+    QUARANTINE_SCRIPT = "#{HOMEBREW_LIBRARY_PATH}/cask/lib/hbc/utils/quarantine.swift".freeze
+
+    # @private
+    def swift
+      @swift ||= DevelopmentTools.locate("swift")
+    end
+
+    def available?
+      status = !swift.nil?
+      odebug "Quarantine is #{status ? "available" : "not available"}."
+      !swift.nil?
+    end
+
+    def detect(file)
+      odebug "Verifying Gatekeeper status of #{file}"
+
+      quarantine_status = !Quarantine.status(file).empty?
+
+      odebug "#{file} is #{quarantine_status ? "quarantined" : "not quarantined"}"
+
+      quarantine_status
+    end
+
+    def status(file, command = SystemCommand)
+      command.run("/usr/bin/xattr",
+                  args:        ["-p", QUARANTINE_ATTRIBUTE, file],
+                  print_stderr: false).stdout
+    end
+
+    def cask(cask, downloaded_path, command = SystemCommand)
+      odebug "Quarantining #{downloaded_path}"
+      quarantiner = command.run(swift,
+                                args: [
+                                  QUARANTINE_SCRIPT,
+                                  downloaded_path,
+                                  cask.url.to_s,
+                                  cask.homepage.to_s,
+                                ])
+
+      return if quarantiner.success?
+
+      raise CaskError, "when quarantining #{downloaded_path}: #{quarantiner.stderr}"
+    end
+
+    def all(downloaded_path, base_path, command = SystemCommand)
+      odebug "Propagating quarantine status of #{downloaded_path}"
+
+      quarantine_status = Quarantine.status(downloaded_path)
+
+      list_of_artifacts = Dir.glob(base_path/"**/*")
```

Comment:
For an application, this will actually loop through each and every file it contains, and not just add the attribute to the enclosing `.app` bundle.

Reply:
What you say is correct; although Gatekeeper only needs `com.apple.quarantine` on the app bundle, manually installing an app from the Finder quarantines the whole directory tree. This replicates that behavior.

Reply:
It's just very inefficient to shell out to `xattr` for every file individually. `xattr` seems to support multiple files as arguments. There are some similar things where we use `xargs` to split the arguments.